compositor_pipeline/src/pipeline.rs

```diff
@@ -39,9 +46,56 @@ pub trait PipelineInput: Send + Sync + 'static {
     fn new(queue: Arc<Queue>, opts: Self::Opts) -> Self;
 }
 
-pub struct Pipeline<Input: PipelineInput, Output: PipelineOutput> {
+pub struct OutputOptions<Receiver: PipelineOutputReceiver> {
+    pub receiver_options: Receiver::Opts,
+    pub encoder_settings: EncoderSettings,
+    pub resolution: Resolution,
+}
+
+pub struct PipelineOutput<Receiver: PipelineOutputReceiver> {
+    sender: Sender<Frame>,
+    receiver_identifier: Receiver::Identifier,
+}
+
+impl<Receiver: PipelineOutputReceiver> PipelineOutput<Receiver> {
+    fn new(opts: OutputOptions<Receiver>) -> Result<Self, String> {
+        let mut encoder = Encoder::new(opts.encoder_settings, opts.resolution)?;
+        let (tx, rx) = crossbeam_channel::unbounded();
+        let (id_tx, id_rx) = crossbeam_channel::bounded(1);
+
+        std::thread::spawn(move || {
+            let mut receiver = Receiver::new(opts.receiver_options, encoder.codec());
+            id_tx.send(receiver.identifier()).unwrap();
+            for frame in rx.iter() {
+                if rx.len() > 20 {
+                    warn!("Dropping frame: encoder queue is too long.");
+                    continue;
+                }
+
+                for packet in encoder.send_frame(frame) {
+                    receiver.send_packet(packet);
+                }
+            }
+        });
+
+        Ok(Self {
+            sender: tx,
+            receiver_identifier: id_rx.recv().unwrap(),
+        })
+    }
+
+    fn send_frame(&self, frame: Frame) {
+        self.sender.send(frame).unwrap();
+    }
+
+    pub fn identifier(&self) -> &Receiver::Identifier {
```

Comment:
I don't think an identifier is a good name. Yes it's correct, but not really what is supporse to mean. I think that instead of using identifier you should use what is currently named PipelineOutputReceiver.

Reply:
I don't think I get what you mean